src/expr/src/expr/expr_array_remove.rs

```diff
@@ -0,0 +1,165 @@
+            .field("left", &self.left)
+            .field("right", &self.right)
+            .finish()
+    }
+}
+
+impl ArrayRemoveExpression {
+    fn new(return_type: DataType, left: BoxedExpression, right: BoxedExpression) -> Self {
+        Self {
+            return_type,
+            left,
+            right,
+        }
+    }
+
+    /// Removes all elements equal to the given value from the array.
+    /// Note the behavior is slightly different from PG.
+    ///
+    /// Examples:
+    ///
+    /// ```slt
+    /// query T
+    /// select array_remove(array[array[1],array[2],array[3],array[2],null::int[]], array[1]);
+    /// ----
+    /// {{2},{3},{2},NULL}
+    ///
+    /// query T
+    /// select array_remove(array[array[1],array[2],array[3],array[2],null::int[]], array[2]);
+    /// ----
+    /// {{1},{3},NULL}
+    ///
+    /// query T
+    /// select array_remove(array[array[1],array[2],array[3],array[2],null::int[]], null::int[]);
+    /// ----
+    /// {{1},{2},{3},{2}}
+    ///
+    /// query T
+    /// select array_remove(array[array[1],array[2],array[3],array[2],null::int[]], array[4]);
+    /// ----
+    /// {{1},{2},{3},{2},NULL}
+    ///
+    /// query T
+    /// select array_remove(null::int[], 1);
+    /// ----
+    /// NULL
+    ///
+    /// query T
+    /// select array_remove(ARRAY[array[1],array[2],array[3],array[2],null::int[]], array[3.14]);
+    /// ----
+    /// {{1},{2},{3},{2},NULL}
+    ///
+    /// statement error
+    /// select array_remove(array[array[1],array[2],array[3],array[2],null::int[]], 1);
+    ///
+    /// statement error
+    /// select array_remove(array[array[1],array[2],array[3],array[2],null::int[]], array[array[3]]);
+    ///
+    /// statement error
+    /// select array_remove(ARRAY[array[1],array[2],array[3],array[2],null::int[]], array[true]);
+    /// ```
+    fn array_remove(left: DatumRef<'_>, right: DatumRef<'_>) -> Datum {
+        match left {
+            Some(ScalarRefImpl::List(left)) => Some(
+                ListValue::new(
+                    left.values_ref()
+                        .into_iter()
+                        .filter(|x| x != &right)
+                        .map(|x| x.map(ScalarRefImpl::into_scalar_impl))
+                        .collect(),
+                )
+                .into(),
+            ),
+            _ => None,
+        }
+    }
+
+    fn evaluate(&self, left: DatumRef<'_>, right: DatumRef<'_>) -> Datum {
+        Self::array_remove(left, right)
+    }
+}
+
+#[async_trait::async_trait]
+impl Expression for ArrayRemoveExpression {
+    fn return_type(&self) -> DataType {
+        self.return_type.clone()
+    }
+
+    async fn eval(&self, input: &DataChunk) -> Result<ArrayRef> {
+        let left_array = self.left.eval_checked(input).await?;
+        let right_array = self.right.eval_checked(input).await?;
+        let mut builder = self
+            .return_type
+            .create_array_builder(left_array.len() + right_array.len());
```

Comment:
This should be `input.capacity()` which should always be the same as `left_array.len()` and `right_array.len()`. We are wasting 1x reserved memory here. (And yes, `array_concat` also looks buggy here...)